src/RamTransforms.cpp

```diff
@@ -641,84 +641,95 @@ bool TupleIdTransformer::reorderOperations(RamProgram& program) {
 bool HoistAggregateTransformer::hoistAggregate(RamProgram& program) {
     bool changed = false;
 
-    // hoist a single aggregate to an outer scope
-    // most outer permissible aggregate is hoisted
+    // There are two cases: aggregates that have no data-dependencies on
+    // other RAM operations, and aggregates that have data-dependencies.
+
+    // Hoist a single aggregate to an outer scope that
+    // has no data-dependency.
+    visitDepthFirst(program, [&](const RamQuery& query) {
+        std::unique_ptr<RamNestedOperation> newAgg;
+        int lastTupleOp = -1;
+
+        // The rewriter has two tasks: (1) identify a single aggregate that
+        // can be hoisted and (2) insert it at the outermost level.
+        // We assume all RamOperations are renumbered for this transformation.
+        std::function<std::unique_ptr<RamNode>(std::unique_ptr<RamNode>)> aggRewriter =
+                [&](std::unique_ptr<RamNode> node) -> std::unique_ptr<RamNode> {
+            // identify hoistable aggregates
+            if (nullptr != dynamic_cast<RamAggregate*>(node.get())) {
+                RamTupleOperation* tupleOp = dynamic_cast<RamTupleOperation*>(node.get());
+                assert(tupleOp != nullptr && "aggregate conversion to tuple operation failed");
+                if (rla->getLevel(tupleOp) == -1 && lastTupleOp != -1) {
+                    changed = true;
+                    newAgg = std::unique_ptr<RamNestedOperation>(
+                            dynamic_cast<RamNestedOperation*>(tupleOp->clone()));
+                    assert(newAgg != nullptr && "failed to make a clone");
+                    return std::unique_ptr<RamOperation>(tupleOp->getOperation().clone());
+                }
+            } else if (RamTupleOperation* tupleOp = dynamic_cast<RamTupleOperation*>(node.get())) {
+                // tuple operation that is a non-aggregate
+                lastTupleOp = tupleOp->getTupleId();
+            }
+            node->apply(makeLambdaRamMapper(aggRewriter));
+            return node;
+        };
+        const_cast<RamQuery*>(&query)->apply(makeLambdaRamMapper(aggRewriter));
+        if (newAgg != nullptr) {
+            newAgg->rewrite(
+                    &newAgg->getOperation(), std::unique_ptr<RamOperation>(query.getOperation().clone()));
+            const_cast<RamQuery*>(&query)->rewrite(&query.getOperation(), std::move(newAgg));
+        }
+    });
+
+    // hoist a single aggregate to an outer scope that
+    // is data-dependent on an operation, i.e., its data dependence level
+    // is not -1 (=independent).
     visitDepthFirst(program, [&](const RamQuery& query) {
         // new level of aggregate
         int newLevel = -1;
-
         // new aggregate
         std::unique_ptr<RamNestedOperation> newAgg;
-
         // level of last non-aggregate
-        int lastNonAggLevel = -1;
+        int lastTupleOp = -1;
```

Comment:
Is this an identifier or a level? It's set from getTupleId, but then compared to levels. and the comment says it's a level. Either the name or the comment (or the code?) should be changed.

Reply:
It is an identifier / level at the same time because we assume a prior renumbering of identifiers.